src/metabase/util/ssh.clj

```diff
@@ -1,32 +1,63 @@
 (ns metabase.util.ssh
-  (:require [clojure.string :as str]
-            [clojure.tools.logging :as log]
+  (:require [clojure.tools.logging :as log]
             [metabase.util :as u])
-  (:import com.jcraft.jsch.JSch))
+  (:import java.io.ByteArrayInputStream
+           [org.apache.sshd.client.future AuthFuture ConnectFuture]
+           org.apache.sshd.client.session.ClientSession
+           org.apache.sshd.client.session.forward.PortForwardingTracker
+           org.apache.sshd.client.SshClient
+           [org.apache.sshd.common.config.keys FilePasswordProvider FilePasswordProvider$ResourceDecodeResult]
+           org.apache.sshd.common.session.SessionHolder
+           org.apache.sshd.common.util.GenericUtils
+           org.apache.sshd.common.util.io.resource.AbstractIoResource
+           org.apache.sshd.common.util.net.SshdSocketAddress
+           org.apache.sshd.common.util.security.SecurityUtils
+           org.apache.sshd.server.forward.AcceptAllForwardingFilter))
 
 (def ^:private default-ssh-timeout 30000)
 
+(def ^:private client
+  (doto (SshClient/setUpDefaultClient)
+    (.start)
+    (.setForwardingFilter AcceptAllForwardingFilter/INSTANCE)))
+
 (defn start-ssh-tunnel
   "Opens a new ssh tunnel and returns the connection along with the dynamically
-   assigned tunnel entrance port. It's the callers responsibility to call .disconnect
+   assigned tunnel entrance port. It's the callers responsibility to call `close-tunnel`
    on the returned connection object."
-  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass host port]}]
-  (let [connection (doto ^com.jcraft.jsch.Session (.getSession (new com.jcraft.jsch.JSch)
-                                                               ^String tunnel-user
-                                                               ^String tunnel-host
-                                                               tunnel-port)
-                     (.setPassword ^String tunnel-pass)
-                     (.setConfig "StrictHostKeyChecking" "no")
-                     (.connect default-ssh-timeout)
-                     (.setPortForwardingL 0 host port))
-        input-port (some-> (.getPortForwardingL connection)
-                           first
-                           (str/split #":")
-                           first
-                           (Integer/parseInt))]
-    (assert (number? input-port))
-    (log/info (u/format-color 'cyan "creating ssh tunnel %s@%s:%s -L %s:%s:%s" tunnel-user tunnel-host tunnel-port input-port host port))
-    [connection input-port]))
+  [{:keys [tunnel-host tunnel-port tunnel-user tunnel-pass tunnel-private-key tunnel-private-key-passphrase host port]}]
+  (let [conn-future (.connect ^SshClient client ^String tunnel-user ^String tunnel-host ^Integer tunnel-port)
+        conn-status (.verify ^ConnectFuture conn-future default-ssh-timeout)
+        session (.getSession ^SessionHolder conn-status)
```

Comment:
wouldn't it be easier to tag `^ClientSession session` here when you bind it once instead of doing it every time you use it below?

Reply:
oh probably true. Didn't realize you could do that.